fixes #2263 search Hostgroup by label

isratrade opened the pull request:
@abenari, @ohadlevy

I made the search param by label.  However, the UI column header is "name".  Do you suggest to change the column to "label" or change the scoped_search on "name" to look at "label" column?

Commit:
fixes #2263 search Hostgroup by label

ohadlevy commented:
can you please add tests?

isratrade commented:
@ohadlevy re-committed with tests

ohadlevy commented:
@isratrade the tests fail, can you have a look please?

isratrade commented:
@ohadlevy the tests pass for Ruby 1.9.2 and 1.9.3, but fail for 1.8.7 on the STI Host method missing.  @GregSutcliffe

Any suggestion how to proceed?

  1) Error:
test_user_with_destroy_permissions_should_be_able_to_destroy(HostgroupTest):
ActiveRecord::AssociationTypeMismatch: Host(#70149929678160) expected, got Host::Managed(#70149926047200)
    app/models/host.rb:11:in `send'
    app/models/host.rb:11:in`method_missing'
    test/unit/hostgroup_test.rb:59:in `test_user_with_destroy_permissions_should_be_able_to_destroy'
    /test/test_helper.rb:31:in`as_user'
    /test/test_helper.rb:37:in `as_admin'
    test/unit/hostgroup_test.rb:58:in`test_user_with_destroy_permissions_should_be_able_to_destroy'

  2) Error:
test_user_with_destroy_permissions_should_not_be_able_to_edit(HostgroupTest):
ActiveRecord::AssociationTypeMismatch: Host(#70149929678160) expected, got Host::Managed(#70149926047200)
    app/models/host.rb:11:in `send'
    app/models/host.rb:11:in`method_missing'
    test/unit/hostgroup_test.rb:84:in `test_user_with_destroy_permissions_should_not_be_able_to_edit'
    /test/test_helper.rb:31:in`as_user'
    /test/test_helper.rb:37:in `as_admin'
    test/unit/hostgroup_test.rb:83:in`test_user_with_destroy_permissions_should_not_be_able_to_edit'

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -124,6 +125,35 @@ def root_pass
+    if name_changed? || ancestry_changed?
+      Hostgroup.where("ancestry IS NOT NULL").each do |hostgroup|
+        if hostgroup.path_ids.include?(self.id)
+          lbl = hostgroup.ancestors.map{|a| a.name + "/"}.join + hostgroup.name
```
shouldnt we use hostgroup.get_label instead?

ohadlevy commented on app/models/hostgroup.rb:
```diff
@@ -124,6 +125,35 @@ def root_pass
+      if ancestry.empty?
+        self.label = name
+      else
+        self.label = ancestors.map{|a| a.name + "/"}.join + name
```
since we use it multiple times, can we extract it into a method?

ohadlevy commented on db/migrate/20130228084503_add_label_to_hostgroup.rb:
```diff
@@ -0,0 +1,21 @@
+class AddLabelToHostgroup < ActiveRecord::Migration
+
+  class Hostgroup < ActiveRecord::Base
+      has_ancestry :orphan_strategy => :rootify
```
any idea why we need this? I thought this is the default?

isratrade replied:
@ohadlevy, Default is :destroy.  From the docs

:orphan_strategy       Instruct Ancestry what to do with children of a node that is destroyed:
                       :destroy   All children are destroyed as well (default)
                       :rootify   The children of the destroyed node become root nodes

ohadlevy commented:
can you rebase as well? thanks

ohadlevy commented:
[test]

isratrade commented:
@ohadlevy rebased

ohadlevy commented:
@isratrade can you have a look at the test failures?

isratrade commented:
@ohadlevy , it passed on postgres and failed on mysql and sqlite for a destroy action ????

isratrade commented:
@ohadlevy, error msg is 
Host(#70298320242100) expected, got Host::Managed(#70298316937260)

I'm guessing my recent commit will pass the tests on mysql and sqlite.  If so, I will squash the 2 commits.

isratrade commented:
@domcleal I'm stumped at the failure on MySql and Sqlite.  Can you confirm that this test databases are migrated.
Host(#70295104949920) expected, got Host::Managed(#70295097400200) (ActiveRecord::AssociationTypeMismatch)

ohadlevy commented:
@isratrade i think its a simple has_many :hosts which is wrong on the hostgroup... probably leftover from the hosts STI migration?

Commit:
continuation of fixes #2254 - STI Host relationships

isratrade commented:
@ohadlevy , @domcleal, @skottler , any ideas why there is one failing test on Jenkins for Ruby 1.8.7 and sqlite3 TEST-TFTPOrchestrationTest.xml

isratrade commented:
@ohadlevy , @skottler , strange that it now fails on Ruby 1.8.7 and Postgres with a completely different error ???
Validation failed: Ip Does not match selected Subnet, Mac can't be blank, Mac is invalid

isratrade commented:
@skottler @domcleal, http://ci.theforeman.org/job/test_develop_pull_request/258/ shows red on Ruby 1.9.2 and sqlite3, but it doesn't give the details of any failed tests???

GregSutcliffe commented:
@isratrade Looks like a network failure for rubygems.org:

```
Fetching gem metadata from http://rubygems.org/...
Fetching source index from http://rubygems.org/
Network error while fetching
```

skottler commented:
Yeah, it was probably just a transient AWS network issue. I'm fine with going ahead with the merge without a retest since the cause was fairly obvious and it only happened on one platform.

isratrade commented:
@domcleal, @skottler, I added :destroy to the callbacks and a test and squashed the commits, but I don't see that Jenkins is running the tests of the new commit.

``` ruby
before_save :set_label, :on => [:create, :update, :destroy]
after_save :set_other_labels, :on => [:update, :destroy]
```

domcleal commented:
@isratrade yeah, CI's down at the moment, investigating.

isratrade commented:
@ohadlevy, @domcleal I updated the migration to use hostgroup.get_label

dLobatog commented:
Could we add hostgroup search (by label) to the Reports model? It's rather useful and doing this without the label column is just a mess.

isratrade commented:
@eLobato sure, I can add this next week in a separate pull request

ohadlevy commented:
merged - thanks!
